base/applications/mspaint/mouse.cpp

```diff
@@ -277,293 +508,374 @@ endPaintingL(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            roundTo8Directions(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_BEZIER
+struct BezierTool : ToolBase
+{
+    BezierTool() : ToolBase(TOOL_BEZIER)
+    {
+    }
+
+    void draw(BOOL bLeftButton)
+    {
+        COLORREF rgb = (bLeftButton ? m_fg : m_bg);
+        switch (pointSP)
+        {
+            case 1:
+                Line(m_hdc, pointStack[0].x, pointStack[0].y, pointStack[1].x, pointStack[1].y, rgb,
+                     toolsModel.GetLineWidth());
+                break;
+            case 2:
+                Bezier(m_hdc, pointStack[0], pointStack[2], pointStack[2], pointStack[1], rgb, toolsModel.GetLineWidth());
+                break;
+            case 3:
+                Bezier(m_hdc, pointStack[0], pointStack[2], pointStack[3], pointStack[1], rgb, toolsModel.GetLineWidth());
+                break;
+        }
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        pointStack[pointSP].x = x;
+        pointStack[pointSP].y = y;
+        if (pointSP == 0)
+        {
             imageModel.CopyPrevious();
-            Airbrush(hdc, x, y, bg, toolsModel.GetAirBrushWidth());
-            break;
-        case TOOL_BEZIER:
-            pointStack[pointSP].x = x;
-            pointStack[pointSP].y = y;
-            if (pointSP == 0)
-            {
-                imageModel.CopyPrevious();
-                pointSP++;
-            }
-            break;
-        case TOOL_SHAPE:
-            pointStack[pointSP].x = x;
-            pointStack[pointSP].y = y;
-            if (pointSP + 1 >= 2)
-                Poly(hdc, pointStack, pointSP + 1, bg, fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle(), FALSE, FALSE);
-            if (pointSP == 0)
-            {
-                imageModel.CopyPrevious();
-                pointSP++;
-            }
-            break;
-        case TOOL_RECTSEL:
-        case TOOL_COLOR:
-        case TOOL_ZOOM:
-            break;
+            pointSP++;
+        }
     }
-}
 
-void
-whilePaintingR(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        pointStack[pointSP].x = x;
+        pointStack[pointSP].y = y;
+        draw(bLeftButton);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        draw(bLeftButton);
```

Comment:
Is the `draw` call needed?

Reply:
Yep. If no drawing, then the line will vanish on mouse button release.